### PR TITLE
Surface tree conflicts.

### LINKED ISSUE #2789: Surface tree conflicts when updating against Github.

### PR DESCRIPTION
**Pull request by seanparsons**
Fixes #2789

**Problem:**
Currently when updating the project against Github it's possible for tree conflicts to occur, but we completely ignore them.

**Fix:**
The previous merge conflict identifying code has now been extended to identify tree conflicts, recording their results in a separate structure so they can be resolved later and potentially to see either side of the conflict.

**Limitations:**
- This does not currently provide any functionality to resolve those conflicts.

**Commit Details:**
- Separated out some arbitrary instances into `assets.test-utils.ts` so they can be imported from elsewhere.
- `deriveGithubFileChanges` now takes account of any tree conflicts which have been passed in.
- `mergeProjectContents` now takes the current time so that changes are against a single time and also to make testing easier.
- `UPDATE_AGAINST_GITHUB` action now takes account of any tree conflicts that have arisen, potentially not updating the origin commit.
- Added `GithubData.treeConflicts` field.
- `mergeProjectContentsTree` reworked to capture tree conflicts.
- Added `Conflict` sum type to capture the different types of conflict which can arise.
- Added `GithubFileChanges.conflicted` field.

### CODE REVIEW

**Review by bkrmendy — APPROVED** [review]

File: editor/src/components/assets.ts [hunk 1]
```diff
@@ -76,25 +77,34 @@ export function deriveGithubFileChanges(
   let untracked: Array<string> = []
   let modified: Array<string> = []
   let deleted: Array<string> = []
+  let conflicted: Array<string> = []
+
+  conflicted = Object.keys(treeConflicts)
```
Comment (Rheeseyb) [inline editor/src/components/assets.ts:82]:
```suggestion
  const conflicted: Array<string> = Object.keys(treeConflicts)
```

File: editor/src/core/shared/github.spec.ts [hunk 2]
```diff
@@ -62,3 +79,372 @@ if (root != null) {
...
+  })
+  it('should return the current contents when a change only exists for that', () => {
+    const originContents = projectContentFile(
+      '/myfile.txt',
+      textFile(
+        textFileContents('origin code', unparsed, RevisionsState.CodeAhead),
+        null,
+        null,
+        1000000,
+      ),
```
Comment (Rheeseyb) [inline editor/src/core/shared/github.spec.ts:116]: You can use `codeFile` to simplify this: <https://github.com/concrete-utopia/utopia/blob/e55cb7c6e431f0eb9c01b01709b3dbdd6cd7f397/editor/src/core/shared/project-file-types.ts#L574-L583>

File: editor/src/core/shared/github.ts [hunk 3]
```diff
@@ -625,63 +765,138 @@ export function mergeProjectContents(
...
     isProjectContentFile(branchContents) &&
     isTextFile(branchContents.content)
   ) {
-    // All 3 branches are a file.
-    const mergedResult = mergeDiff3(
-      currentContents.content.fileContents.code,
-      originContents.content.fileContents.code,
-      branchContents.content.fileContents.code,
-      {
+    // At this path it was and in Utopia and in the branch is a text file.
```
Comment (Rheeseyb) [inline editor/src/core/shared/github.ts:782]: I think there's a typo here, or a missed word, or something...

Comment (seanparsons) [reply editor/src/core/shared/github.ts]: Fixed in fe952b182.

File: editor/src/core/shared/github.spec.ts (hunk 2 above)
Comment (seanparsons) [reply editor/src/core/shared/github.spec.ts]: Fixed in fe952b182.

**Pull request merged**